File: MobileWeather/LocationService.m
//
//  LocationService.m
//  MobileWeather
//
//  Copyright (c) 2013-2015 Ford Motor Company. All rights reserved.
//

#import <CoreLocation/CoreLocation.h>

#import "Notifications.h"
#import "LocationService.h"
#import "WeatherLocation.h"

NSTimeInterval const MinimumTimeBetweenLocationUpdates = 120;

@interface LocationService () <CLLocationManagerDelegate>

@property (strong, nonatomic) CLLocationManager *manager;
@property (strong, nonatomic) CLGeocoder *geocoder;
@property (strong, nonatomic) NSDate *lastLocationUpdate;

@end

@implementation LocationService

+ (instancetype)sharedService {
    static LocationService *shared = nil;
    static dispatch_once_t token;
    
    dispatch_once(&token, ^{
        shared = [[LocationService alloc] init];
    });
    
    return shared;
}

- (instancetype)init {
    if (self = [super init]) {
        _manager = [[CLLocationManager alloc] init];
        _manager.delegate = self;
        _manager.pausesLocationUpdatesAutomatically = YES;
        // specify accuracy of 500 meters. Its enough for determining the city and will safe battery.
        _manager.desiredAccuracy = 500.0;
        // Set the minimum distance that the device must move before we want an update
        _manager.distanceFilter = 500.0;

        [_manager requestWhenInUseAuthorization];
        
        _geocoder = [[CLGeocoder alloc] init];
    }
    return self;
}

- (void)start {
    self.lastLocationUpdate = [NSDate dateWithTimeIntervalSince1970:0];
    [self.manager startMonitoringSignificantLocationChanges];
}

- (void)stop {
    [self.manager stopMonitoringSignificantLocationChanges];
}

#pragma mark - CLLocationManagerDelegate

- (void)locationManager:(CLLocationManager *)manager didChangeAuthorizationStatus:(CLAuthorizationStatus)status {
    NSString *statusString;

    switch (status) {
        case kCLAuthorizationStatusAuthorizedWhenInUse:
            statusString = @"AuthorizedWhenInUse";
            [self.manager startUpdatingLocation];
            break;
        case kCLAuthorizationStatusDenied:
            statusString = @"Denied";
            break;
        case kCLAuthorizationStatusNotDetermined:
            statusString = @"NotDetermined";
            break;
        case kCLAuthorizationStatusRestricted:
            statusString = @"Restricted";
            break;
        default:
            statusString = @"Unknown";
            break;
    }
}

- (void)locationManager:(CLLocationManager *)manager didFailWithError:(NSError *)error {
    NSLog(@"%s\n%@", __FUNCTION__, error.localizedFailureReason ?: error);
}

- (void)locationManagerDidPauseLocationUpdates:(CLLocationManager *)manager {
    NSLog(@"Location Manager did pause location updates");
}

- (void)locationManagerDidResumeLocationUpdates:(CLLocationManager *)manager {
    NSLog(@"Location Manager did resume location updates");
}

- (void)locationManager:(CLLocationManager *)manager didUpdateLocations:(NSArray<CLLocation *> *)locations {
    NSLog(@"%s\n%@", __FUNCTION__, locations);

    NSDate *date = [NSDate date];
    if ([date timeIntervalSinceDate:self.lastLocationUpdate] <= MinimumTimeBetweenLocationUpdates) {
        return;
    }

    self.lastLocationUpdate = date;
    [self.geocoder reverseGeocodeLocation:locations.lastObject completionHandler:^(NSArray *placemarks, NSError *error) {
        if (placemarks.count == 0) { return; }

        CLPlacemark *placemark = placemarks.lastObject;
        WeatherLocation *location = [[WeatherLocation alloc] initWithCity:placemark.locality state:placemark.administrativeArea zipCode:placemark.postalCode country:placemark.country gpsLocation:placemark.location];

        [[NSNotificationCenter defaultCenter] postNotificationName:MobileWeatherLocationUpdateNotification object:self userInfo:@{ @"location": location }];
    }];
}

@end

File: MobileWeather/LocationService.swift
//
//  LocationService.swift
//  WeatherLocationService
//
//  Created by Joel Fischer on 6/7/21.
//  Copyright © 2021 Ford. All rights reserved.
//

import CoreLocation
import Foundation

protocol LocationServiceDelegate: AnyObject {
    func locationDidUpdate(newLocation: WeatherLocation)
}

class LocationService: NSObject {
    private let locationManager = CLLocationManager()
    private let geocoder = CLGeocoder()
    private var lastLocationUpdate = Date(timeIntervalSince1970: 0)
    private let minimumLocationUpdateSeconds = 120.0
    private weak var delegate: LocationServiceDelegate?

    public init(delegate: LocationServiceDelegate) {
        self.delegate = delegate

        locationManager.pausesLocationUpdatesAutomatically = true
        locationManager.desiredAccuracy = 500.0
        locationManager.distanceFilter = 500.0
        locationManager.requestWhenInUseAuthorization()

        super.init()
        locationManager.delegate = self
    }

    @objc public func start() {
        locationManager.startMonitoringSignificantLocationChanges()
    }

    @objc public func stop() {
        locationManager.stopMonitoringSignificantLocationChanges()
    }
}

extension LocationService: CLLocationManagerDelegate {
    public func locationManager(_ manager: CLLocationManager, didChangeAuthorization status: CLAuthorizationStatus) {
        switch status {
        case .authorizedWhenInUse: fallthrough
        case .authorizedAlways:
            manager.startUpdatingLocation()
        case .notDetermined: break
        case .denied: fallthrough
        case .restricted: fallthrough
        default:
            // TODO: Something to display that location isn't allowed
            break;
        }
    }

    public func locationManager(_ manager: CLLocationManager, didUpdateLocations locations: [CLLocation]) {
        let now = Date()
        guard locations.count > 0,
              now.timeIntervalSince(lastLocationUpdate) > minimumLocationUpdateSeconds else { return }

        lastLocationUpdate = now
        geocoder.reverseGeocodeLocation(locations.last!) { placemarks, error in
            guard let place = placemarks?.last, let gpsLocation = place.location else {
                print("Error reverse geocoding location: \(locations.last!), error: \(error!)")
                return
            }

            let weatherLocation = WeatherLocation(country: place.country, state: place.administrativeArea, city: place.locality, zipCode: place.postalCode, gpsLocation: gpsLocation)
            self.delegate?.locationDidUpdate(newLocation: weatherLocation)
        }
    }
}
